packages/services/src/activity/activity.service.ts

```diff
@@ -0,0 +1,335 @@
+  }
+
+  async function getStacksTxActivity(
+    tx: HiroStacksTransaction | HiroStacksMempoolTransaction,
+    txEvents: HiroTransactionEvent[] = [],
+    account: AccountIdentifier,
+    signal?: AbortSignal
+  ): Promise<OnChainActivity[]> {
+    switch (tx.tx_type) {
+      case 'token_transfer':
+        return mapTokenTransferActivity(tx, account);
+      case 'smart_contract':
+        return mapSmartContractActivity(
+          tx,
+          account,
+          await getStacksAssetTransfersWithInfo(tx, txEvents, signal)
+        );
+      case 'contract_call':
+        return mapContractCallActivity(
+          tx,
+          account,
+          await getStacksAssetTransfersWithInfo(tx, txEvents, signal)
+        );
+      default:
+        return [];
+    }
+  }
+
+  async function getStacksAssetTransfersWithInfo(
+    tx: HiroStacksTransaction | HiroStacksMempoolTransaction,
+    txEvents: HiroTransactionEvent[],
+    signal?: AbortSignal
+  ): Promise<StacksAssetTransferWithInfo[]> {
+    const transfers = getStacksAssetTransfers(tx, txEvents);
+    return (await Promise.all(transfers.map(t => addStacksTransferAssetInfo(t, signal)))).filter(
+      isDefined
+    );
+  }
+
+  async function addStacksTransferAssetInfo(
+    transfer: StacksAssetTransfer,
+    signal?: AbortSignal
+  ): Promise<StacksAssetTransferWithInfo | undefined> {
+    if (transfer.assetId === 'STX') {
+      return {
+        ...transfer,
+        assetInfo: stxCryptoAsset,
+      };
+    } else {
+      try {
+        return {
+          ...transfer,
+          assetInfo:
+            transfer.assetCategory === CryptoAssetCategories.fungible
+              ? await sip10AssetService.getAssetInfo(transfer.assetId, signal)
+              : await sip9AssetService.getAssetInfo(transfer.assetId, transfer.tokenValue!, signal),
+        };
+      } catch {
+        // if we're unable to find asset metadata, remove transfer
+        return undefined;
+      }
+    }
+  }
+
+  async function applyMarketData(activity: Activity, signal?: AbortSignal) {
+    if (
+      (activity.type === OnChainActivityTypes.sendAsset ||
+        activity.type === OnChainActivityTypes.receiveAsset) &&
+      activity.asset.category === CryptoAssetCategories.fungible
+    ) {
+      const assetMarketData = await marketDataService.getMarketData(activity.asset, signal);
+      const cryptoValue = createMoney(
+        activity.amount,
+        activity.asset.symbol,
+        activity.asset.decimals
+      );
+      return {
+        ...activity,
+        value: {
+          crypto: cryptoValue,
+          fiat: baseCurrencyAmountInQuote(cryptoValue, assetMarketData),
+        },
+      };
+    } else if (
+      activity.type === OnChainActivityTypes.swapAssets &&
+      (activity.fromAsset.category === CryptoAssetCategories.fungible ||
+        activity.toAsset.category === CryptoAssetCategories.fungible)
+    ) {
+      let fromValue,
```

Comment:
Maybe you don't need to use `let` here? 

You can return `undefined` at the end instead:
```
else {
      return { 
        ...activity,
        fromValue: undefined ,
        toValue: undefined,
    }
```